File: src/packages/llm_gateway/stub.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass
import json
import os
from typing import AsyncIterator
import uuid

from packages.config import load_dotenv_if_enabled

from .contract import (
    LlmGatewayRequest,
    LlmGatewayStreamEvent,
    LlmStreamLlmRequest,
    LlmStreamLlmResponseChunk,
    LlmStreamMessageDelta,
    LlmStreamRunCompleted,
)
from .gateway import LlmGateway

_STUB_ENABLED_ENV = "ARKLOOP_STUB_AGENT_ENABLED"
_STUB_DELTA_COUNT_ENV = "ARKLOOP_STUB_AGENT_DELTA_COUNT"
_STUB_DELTA_INTERVAL_SECONDS_ENV = "ARKLOOP_STUB_AGENT_DELTA_INTERVAL_SECONDS"
_LLM_DEBUG_EVENTS_ENV = "ARKLOOP_LLM_DEBUG_EVENTS"

_DEFAULT_STUB_ENABLED = True
_DEFAULT_STUB_DELTA_COUNT = 3
_DEFAULT_STUB_DELTA_INTERVAL_SECONDS = 0.02


def _parse_bool(value: str) -> bool:
    cleaned = value.strip().casefold()
    if cleaned in {"1", "true", "yes", "y", "on"}:
        return True
    if cleaned in {"0", "false", "no", "n", "off"}:
        return False
    raise ValueError("必须为布尔值（0/1、true/false）")


def _parse_positive_int(value: str) -> int:
    cleaned = value.strip()
    parsed = int(cleaned)
    if parsed <= 0:
        raise ValueError("必须为正整数")
    return parsed


def _parse_non_negative_float(value: str) -> float:
    cleaned = value.strip()
    parsed = float(cleaned)
    if parsed < 0:
        raise ValueError("必须为非负数")
    return parsed


@dataclass(frozen=True, slots=True)
class StubLlmGatewayConfig:
    enabled: bool = _DEFAULT_STUB_ENABLED
    delta_count: int = _DEFAULT_STUB_DELTA_COUNT
    delta_interval_seconds: float = _DEFAULT_STUB_DELTA_INTERVAL_SECONDS
    emit_llm_debug_events: bool = False

    @classmethod
    def from_env(cls) -> "StubLlmGatewayConfig":
        load_dotenv_if_enabled(override=False)

        enabled = _DEFAULT_STUB_ENABLED
        raw_enabled = os.getenv(_STUB_ENABLED_ENV)
        if raw_enabled:
            enabled = _parse_bool(raw_enabled)

        delta_count = _DEFAULT_STUB_DELTA_COUNT
        raw_delta_count = os.getenv(_STUB_DELTA_COUNT_ENV)
        if raw_delta_count:
            delta_count = _parse_positive_int(raw_delta_count)

        delta_interval_seconds = _DEFAULT_STUB_DELTA_INTERVAL_SECONDS
        raw_delta_interval_seconds = os.getenv(_STUB_DELTA_INTERVAL_SECONDS_ENV)
        if raw_delta_interval_seconds:
            delta_interval_seconds = _parse_non_negative_float(raw_delta_interval_seconds)

        emit_llm_debug_events = False
        raw_debug = os.getenv(_LLM_DEBUG_EVENTS_ENV)
        if raw_debug:
            emit_llm_debug_events = _parse_bool(raw_debug)

        return cls(
            enabled=enabled,
            delta_count=delta_count,
            delta_interval_seconds=delta_interval_seconds,
            emit_llm_debug_events=emit_llm_debug_events,
        )


class StubLlmGateway(LlmGateway):
    def __init__(self, *, config: StubLlmGatewayConfig) -> None:
        self._config = config

    async def stream(self, *, request: LlmGatewayRequest) -> AsyncIterator[LlmGatewayStreamEvent]:
        llm_call_id = uuid.uuid4().hex
        if self._config.emit_llm_debug_events:
            yield LlmStreamLlmRequest(
                llm_call_id=llm_call_id,
                provider_kind="stub",
                api_mode="stub",
                payload_json=request.to_json(),
            )

        for index in range(1, self._config.delta_count + 1):
            await asyncio.sleep(self._config.delta_interval_seconds)
            delta = f"stub delta {index}"
            if self._config.emit_llm_debug_events:
                chunk_json = {"content_delta": delta, "role": "assistant"}
                yield LlmStreamLlmResponseChunk(
                    llm_call_id=llm_call_id,
                    provider_kind="stub",
                    api_mode="stub",
                    raw=json.dumps(chunk_json, ensure_ascii=False, separators=(",", ":")),
                    chunk_json=chunk_json,
                )
            yield LlmStreamMessageDelta(content_delta=delta, role="assistant")
        yield LlmStreamRunCompleted()


__all__ = ["StubLlmGateway", "StubLlmGatewayConfig"]

File: src/services/worker/internal/llm/stub.go
package llm

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"arkloop/services/worker/internal/stablejson"
	"github.com/google/uuid"
)

const (
	stubEnabledEnv           = "ARKLOOP_STUB_AGENT_ENABLED"
	stubDeltaCountEnv        = "ARKLOOP_STUB_AGENT_DELTA_COUNT"
	stubDeltaIntervalSeconds = "ARKLOOP_STUB_AGENT_DELTA_INTERVAL_SECONDS"
	llmDebugEventsEnv        = "ARKLOOP_LLM_DEBUG_EVENTS"
	defaultStubEnabled       = true
	defaultStubDeltaCount    = 3
	defaultStubDeltaInterval = 0.02
	defaultStubProviderKind  = "stub"
	defaultStubAPIMode       = "stub"
	truthyValues             = "1,true,yes,y,on"
	falsyValues              = "0,false,no,n,off"
)

type AuxGatewayConfig struct {
	Enabled         bool
	DeltaCount      int
	DeltaInterval   time.Duration
	EmitDebugEvents bool
}

func AuxGatewayConfigFromEnv() (AuxGatewayConfig, error) {
	enabled := defaultStubEnabled
	if raw := strings.TrimSpace(os.Getenv(stubEnabledEnv)); raw != "" {
		value, err := parseBool(raw)
		if err != nil {
			return AuxGatewayConfig{}, fmt.Errorf("%s: %w", stubEnabledEnv, err)
		}
		enabled = value
	}

	deltaCount := defaultStubDeltaCount
	if raw := strings.TrimSpace(os.Getenv(stubDeltaCountEnv)); raw != "" {
		value, err := parsePositiveInt(raw)
		if err != nil {
			return AuxGatewayConfig{}, fmt.Errorf("%s: %w", stubDeltaCountEnv, err)
		}
		deltaCount = value
	}

	intervalSeconds := defaultStubDeltaInterval
	if raw := strings.TrimSpace(os.Getenv(stubDeltaIntervalSeconds)); raw != "" {
		value, err := parseNonNegativeFloat(raw)
		if err != nil {
			return AuxGatewayConfig{}, fmt.Errorf("%s: %w", stubDeltaIntervalSeconds, err)
		}
		intervalSeconds = value
	}

	emitDebugEvents := false
	if raw := strings.TrimSpace(os.Getenv(llmDebugEventsEnv)); raw != "" {
		value, err := parseBool(raw)
		if err != nil {
			return AuxGatewayConfig{}, fmt.Errorf("%s: %w", llmDebugEventsEnv, err)
		}
		emitDebugEvents = value
	}

	return AuxGatewayConfig{
		Enabled:         enabled,
		DeltaCount:      deltaCount,
		DeltaInterval:   time.Duration(intervalSeconds * float64(time.Second)),
		EmitDebugEvents: emitDebugEvents,
	}, nil
}

type AuxGateway struct {
	cfg AuxGatewayConfig
}

func NewAuxGateway(cfg AuxGatewayConfig) *AuxGateway {
	return &AuxGateway{cfg: cfg}
}

func (g *AuxGateway) Stream(ctx context.Context, request Request, yield func(StreamEvent) error) error {
	if !g.cfg.Enabled {
		failed := StreamRunFailed{
			Error: GatewayError{
				ErrorClass: ErrorClassProviderNonRetryable,
				Message:    "aux gateway is disabled",
			},
		}
		return yield(failed)
	}

	llmCallID := uuid.NewString()
	stats := ComputeRequestStats(request)
	debugPayload, redactedHints := sanitizeDebugPayloadJSON(request.ToJSON())
	if err := yield(StreamLlmRequest{
		LlmCallID:            llmCallID,
		ProviderKind:         defaultStubProviderKind,
		APIMode:              defaultStubAPIMode,
		PayloadJSON:          debugPayload,
		RedactedHints:        redactedHints,
		SystemBytes:          stats.SystemBytes,
		ToolsBytes:           stats.ToolsBytes,
		MessagesBytes:        stats.MessagesBytes,
		AbstractRequestBytes: stats.AbstractRequestBytes,
		ProviderPayloadBytes: stats.AbstractRequestBytes,
		ImagePartCount:       stats.ImagePartCount,
		Base64ImageBytes:     stats.Base64ImageBytes,
		RoleBytes:            stats.RoleBytes,
		ToolSchemaBytesMap:   stats.ToolSchemaBytesMap,
		StablePrefixHash:     stats.StablePrefixHash,
	}); err != nil {
		return err
	}

	for idx := 1; idx <= g.cfg.DeltaCount; idx++ {
		if err := sleepWithContext(ctx, g.cfg.DeltaInterval); err != nil {
			return err
		}
		delta := fmt.Sprintf("stub delta %d", idx)
		if g.cfg.EmitDebugEvents {
			chunkJSON := map[string]any{"content_delta": delta, "role": "assistant"}
			raw, _ := stablejson.Encode(chunkJSON)
			if raw == "" {
				raw = string(mustJSONMarshal(chunkJSON))
			}
			if err := yield(StreamLlmResponseChunk{
				LlmCallID:    llmCallID,
				ProviderKind: defaultStubProviderKind,
				APIMode:      defaultStubAPIMode,
				Raw:          raw,
				ChunkJSON:    chunkJSON,
			}); err != nil {
				return err
			}
		}
		if err := yield(StreamMessageDelta{ContentDelta: delta, Role: "assistant"}); err != nil {
			return err
		}
	}
	return yield(StreamRunCompleted{LlmCallID: llmCallID})
}

func sleepWithContext(ctx context.Context, d time.Duration) error {
	if d <= 0 {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
			return nil
		}
	}
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-timer.C:
		return nil
	}
}

func parseBool(raw string) (bool, error) {
	cleaned := strings.ToLower(strings.TrimSpace(raw))
	for _, item := range strings.Split(truthyValues, ",") {
		if cleaned == item {
			return true, nil
		}
	}
	for _, item := range strings.Split(falsyValues, ",") {
		if cleaned == item {
			return false, nil
		}
	}
	return false, fmt.Errorf("must be a boolean (0/1, true/false)")
}

func parsePositiveInt(raw string) (int, error) {
	parsed, err := strconv.Atoi(strings.TrimSpace(raw))
	if err != nil {
		return 0, fmt.Errorf("must be a positive integer")
	}
	if parsed <= 0 {
		return 0, fmt.Errorf("must be a positive integer")
	}
	return parsed, nil
}

func parseNonNegativeFloat(raw string) (float64, error) {
	parsed, err := strconv.ParseFloat(strings.TrimSpace(raw), 64)
	if err != nil {
		return 0, fmt.Errorf("must be non-negative")
	}
	if parsed < 0 {
		return 0, fmt.Errorf("must be non-negative")
	}
	return parsed, nil
}

func mustJSONMarshal(value any) []byte {
	encoded, err := json.Marshal(value)
	if err != nil {
		return []byte("{}")
	}
	return encoded
}
